spec.html

```diff
@@ -29088,9 +28954,106 @@ <h1>encodeURIComponent ( _uriComponent_ )</h1>
         <p>The `encodeURIComponent` function is the <dfn>%encodeURIComponent%</dfn> intrinsic object. When the `encodeURIComponent` function is called with one argument _uriComponent_, the following steps are taken:</p>
         <emu-alg>
           1. Let _componentString_ be ? ToString(_uriComponent_).
-          1. Let _unescapedURIComponentSet_ be a String containing one instance of each code unit valid in |uriUnescaped|.
-          1. Return ? Encode(_componentString_, _unescapedURIComponentSet_).
+          1. Let _extraUnescaped_ be the empty String.
+          1. Return ? Encode(_componentString_, _extraUnescaped_).
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-encode" type="abstract operation">
+        <h1>
+          Encode (
+            _string_: a String,
+            _extraUnescaped_: a String,
+          ): either a normal completion containing a String or a throw completion
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>It performs URI encoding and escaping, interpreting _string_ as a sequence of UTF-16 encoded code points as described in <emu-xref href="#sec-ecmascript-language-types-string-type"></emu-xref>. If a character is identified as unreserved in RFC 2396 or appears in _extraUnescaped_, it is not escaped.</dd>
+        </dl>
+        <emu-alg>
+          1. Let _strLen_ be the length of _string_.
+          1. Let _R_ be the empty String.
+          1. Let _alwaysUnescaped_ be the string-concatenation of ASCII word characters and *"-.!~\*'()"*.
+          1. Let _unescapedSet_ be the string-concatenation of _alwaysUnescaped_ and _extraUnescaped_.
+          1. Let _k_ be 0.
+          1. Repeat,
+            1. If _k_ = _strLen_, return _R_.
+            1. Let _C_ be the code unit at index _k_ within _string_.
+            1. If _C_ is in _unescapedSet_, then
+              1. Set _k_ to _k_ + 1.
+              1. Set _R_ to the string-concatenation of _R_ and _C_.
+            1. Else,
+              1. Let _cp_ be CodePointAt(_string_, _k_).
+              1. If _cp_.[[IsUnpairedSurrogate]] is *true*, throw a *URIError* exception.
+              1. Set _k_ to _k_ + _cp_.[[CodeUnitCount]].
+              1. Let _Octets_ be the List of octets resulting by applying the UTF-8 transformation to _cp_.[[CodePoint]].
+              1. For each element _octet_ of _Octets_, do
+                1. Set _R_ to the string-concatenation of:
+                  * _R_
+                  * *"%"*
+                  * the String representation of _octet_, formatted as a two-digit uppercase hexadecimal number, padded to the left with a zero if necessary
```

Comment:
I know this AO is more of a move than a rewrite, but I'd still prefer to see this change to use `StringPad`:

```suggestion
                1. Let _hex_ be the String representation of _octet_, formatted as a two-digit uppercase hexadecimal number.
                1. Set _R_ to the string-concatenation of _R_, *"%"*, and StringPad(_hex_, 2, *"0"*, ~start~).
```

Reply:
Done, and opened #2852 for the instance of "padded to the left" that is outside of the operations being modified by this PR.